frontend_tests/puppeteer_tests/03-compose.js

```diff
@@ -0,0 +1,174 @@
+    await page.waitForSelector("#private_message_recipient", {visible: true});
+    await common.pm_recipient.expect(page, "user8@zulip.testserver");
+
+    // The last message(private) in the narrow is selected.
+    // Now we go up and open compose box with r key
+    await page.keyboard.press("Escape");
+    await page.keyboard.press("KeyK");
+    await page.keyboard.press("KeyR");
+    await common.check_form_contents(page, '#send_message_form', {
+        stream_message_recipient_stream: "Verona",
+        stream_message_recipient_topic: "Reply test",
+        content: "",
+    });
+
+    // Test opening and closing of compose box.
+    await page.waitForSelector('#stream-message', {hidden: false});
+    await close_compose_box(page);
+    await page.waitForSelector('#stream-message', {hidden: true});
+
+    await page.keyboard.press('KeyX');
+    await page.waitForSelector('#private-message', {hidden: false});
+    await close_compose_box(page);
+    await page.waitForSelector('#private-message', {hidden: true});
+
+
+    // Test focus after narrowing to PMs with a user and typing 'c'
+    const you_and_cordelia_selector = '*[title="Narrow to your private messages with Cordelia Lear"]';
+    // For some unknown reason page.click() isn't working here.
+    await page.evaluate((selector) => document.querySelector(selector).click(),
+                        you_and_cordelia_selector);
+    const cordelia_user_id = 8;
+    const pm_list_selector = `li[data-user-ids-string="${cordelia_user_id}"].expanded_private_message.active-sub-filter`;
+    await page.waitForSelector(pm_list_selector, {visible: true});
+    await close_compose_box(page);
+    await page.keyboard.press('KeyC');
+    await page.waitForSelector('#compose', {visible: true});
+    await page.waitForFunction(() => {
+        return document.activeElement === $('.compose_table #stream_message_recipient_stream')[0];
+    });
+    await close_compose_box(page);
+
+    // Make sure multiple PM recipients display properly.
+    const recipients = ['cordelia@zulip.com', 'othello@zulip.com'];
+    const multiple_recipients_pm = 'A huddle to check spaces';
+    const pm_selector = `.messagebox:contains('${multiple_recipients_pm}')`;
+    await common.send_message(page, 'private', {
+        recipient: recipients.join(', '),
+        outside_view: true,
+        content: multiple_recipients_pm,
+    });
+
+    // Go back to all messages view and make sure all messages are loaded.
+    await page.keyboard.press('Escape');
+
+    await page.waitForSelector('#zhome .message_row', {visible: true});
+    await page.waitForFunction((selector) => {
+        return $(selector).length !== 0;
+    }, {}, pm_selector);
+    await page.evaluate((selector) => {
+        $(selector).slice(-1)[0].click();
+    }, pm_selector);
+    await page.waitForSelector('#compose-textarea', {visible: true});
+    await common.pm_recipient.expect(page, 'user12@zulip.testserver,user8@zulip.testserver');
```

Comment:
@timabbott Can we change the "Who can access user email address?" setting to all users and use the normal emails here?

Reply:
I think we could reasonably make `populate_db` turn that setting off when building the database for casper tests, if we think it'll lead to better code here.  But I'd also wonder if we can just be using user IDs more broadly in these sorts of tests; we've been moving that direction in the Node tests.

Reply:
We should definitely use user IDs more broadly in the new tests.  We started to retrofit that a bit in the Casper tests, but probably not in the compose tests that @chdinesh1089 was working off.  We should definitely have little helpers like `get_email('hamlet')`.

I'm not completely opposed to turning off the setting in `populate_db`, but I would be somewhat cautious there.

Reply:
The concerns I had with using the email comprising of user ID was readability and the issues that might occur if user IDs get changed later(This probably will never happen I guess)

Having a function like `get_internal_email('hamlet')` feels like a better solution. That way we can easily test user IDs also.
Could you give some pointers on how to write such a function? I 'm not sure if I know how to do it now.

Reply:
User IDs are immutable within the lifetime of a database, but there is definitely a possibility of IDs changing from test run to test run.  

I believe just doing a `casper.evaluate()` equivalent to ask `people.js` for data should work well.